### PR TITLE
V1.8.1 maglev

### PR DESCRIPTION
**Pull request by lrouter**
Hi, we have implemented maglev hash based on facebook katran project. 
Now I know that cilium will support maglev hash. I think our patch could make some help. Because now only external traffic will be processed by maglev hash by cilium.
The code have been test and run in our environment for a long time.

Thanks.

### CODE REVIEW

**Comment by aanm** [comment]
I've changed the base branch to 1.8 to make it easier to review.

File: bpf/lib/maglev.h [hunk 1]
```diff
@@ -0,0 +1,153 @@
+/* SPDX-License-Identifier: GPL-2.0 */
+/* Copyright (C) 2016-2020 Authors of Cilium */
+
+#ifndef __MAGLEV_H_
+#define __MAGLEV_H_
```
Comment (joestringer) [inline bpf/lib/maglev.h:5]: The PR description implies the ability to apply this maglev logic also to east-west traffic, not just north-south. @borkmann was there any particular motivation for why the current implementation doesn't address this case? Or is it just not the focus of the initially scoped work?

File: bpf/lib/maglev.h [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+
```
Comment (joestringer) [inline bpf/lib/maglev.h:6]: Are you aware of any other features or properties that this implementation provides which are not covered by the current implementation in the `master` branch (from PR #13131)?

**Review by joestringer — COMMENTED** [review]
:exploding_head: It's awesome to see confirmation from users that we're introducing features that are broadly useful. I haven't performed a proper review of this PR given the overlap with https://github.com/cilium/cilium/pull/13131, but I have some questions below for threaded discussion.

Given that the implementation in #13131  was already merged into the `master` branch, the interesting thing to discuss here to me is to figure out the diff compared to this submission, and look to rebase any extensions from this PR against the merged implementation. 

Thanks for opening this for discussion :D

File: bpf/lib/maglev.h (hunk 1 above)
Comment (borkmann) [reply bpf/lib/maglev.h:5]: We didn't add it to the E-W case which uses the socket LB given for TCP and connected UDP, we directly pick a backend _once_ during connect() call which then the kernel uses for the rest of the socket lifetime. The unconnected UDP case could change backends if the order changes, but for that case there is already the session affinity to have the backends be sticky for a user-defined period of time before selecting new ones (https://docs.cilium.io/en/v1.8/gettingstarted/kubeproxy-free/#session-affinity). Either case, the socket LB doesn't go through any intermediate L4LB mechanism (in the sense of a middle node hop) but always directly to the backend, hence we didn't add it there. Is there a rationale for the case where session affinity should not suffice for unconnected UDP?

File: bpf/lib/maglev.h (hunk 2 above)
Comment (borkmann) [reply bpf/lib/maglev.h:6]: Awesome work @lrouter! From a quick glance it looks like the core bits of the implementation are very similar with  #13131 which was merged ~5 days ago. You do have some additional agent bits and even more test cases. Would it be possible to rebase, clean up & integrate these against latest master so we can review the missing bits and get them integrated? That would be really nice. We should also remove the vendor'ed murmur3 and switch to your implementation. Looking forward!

Comment (lrouter) [reply bpf/lib/maglev.h:6]: @borkmann Sure. I will rebase it based on the latest master.

File: bpf/lib/maglev.h (hunk 1 above)
Comment (AirbornWdd) [reply bpf/lib/maglev.h:5]: I checked the latest code, when `backend_id` is 0, just pick up a backend randomly (not maglev way). I didn't see the directly picking routine. Could someone tell me what important routine I missed.
```
	if (backend_id == 0) {
		backend_from_affinity = false;

		key.backend_slot = (sock_select_slot(ctx_full) % svc->count) + 1;
		backend_slot = __lb4_lookup_backend_slot(&key);
		if (!backend_slot) {
			update_metrics(0, METRIC_EGRESS, REASON_LB_NO_BACKEND_SLOT);
			return -ENOENT;
		}

		backend_id = backend_slot->backend_id;
		backend = __lb4_lookup_backend(backend_id);
	}
```
https://github.com/cilium/cilium/blob/master/bpf/bpf_sock.c#L333

Comment (pchaigno) [reply bpf/lib/maglev.h:5]: @AirbornWdd We don't use Maglev in that case. That's the case @borkmann was referring to as `E-W`:

> We didn't add it to the E-W case which uses the socket LB given for TCP and connected UDP, we directly pick a backend once during connect() call which then the kernel uses for the rest of the socket lifetime.

**Comment by brb** [comment]
@lrouter @AirbornWdd Hey, are you still planning to rebase against the Maglev implementation in the master branch? One feature which is missing is backend weights which you have implemented. It'd be great to get it.

**Comment by lrouter** [comment]
> @lrouter @AirbornWdd Hey, are you still planning to rebase against the Maglev implementation in the master branch? One feature which is missing is backend weights which you have implemented. It'd be great to get it.

OK. I will continue to work for it. It affects a lot of code.

**Comment by brb** [comment]
This can be closed in favor of #13943.

**Pull request closed without merging**